src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/Metrics/AggregationManager.cs

```diff
@@ -0,0 +1,312 @@
+
+        public AggregationManager(
+            Action<Instrument, LabeledAggregationStatistics> collectMeasurement,
+            Action<DateTime> beginCollection,
+            Action<DateTime> endCollection,
+            Action<Instrument> beginInstrumentMeasurements,
+            Action<Instrument> endInstrumentMeasurements,
+            Action<Instrument> instrumentPublished,
+            Action initialInstrumentEnumerationComplete,
+            Action<Exception> collectionError)
+        {
+            _collectMeasurement = collectMeasurement;
+            _beginCollection = beginCollection;
+            _endCollection = endCollection;
+            _beginInstrumentMeasurements = beginInstrumentMeasurements;
+            _endInstrumentMeasurements = endInstrumentMeasurements;
+            _instrumentPublished = instrumentPublished;
+            _initialInstrumentEnumerationComplete = initialInstrumentEnumerationComplete;
+            _collectionError = collectionError;
+
+            _listener = new MeterListener()
+            {
+                InstrumentPublished = (instrument, listener) =>
+                {
+                    _instrumentPublished(instrument);
+                    InstrumentState? state = GetInstrumentState(instrument);
+                    if (state != null)
+                    {
+                        _beginInstrumentMeasurements(instrument);
+                        listener.EnableMeasurementEvents(instrument, state);
+                    }
+                },
+                MeasurementsCompleted = (instrument, cookie) =>
+                {
+                    _endInstrumentMeasurements(instrument);
+                    RemoveInstrumentState(instrument, (InstrumentState)cookie!);
+                }
+            };
+            _listener.SetMeasurementEventCallback<double>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<float>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<long>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<int>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<short>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<byte>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+            _listener.SetMeasurementEventCallback<decimal>((i, m, l, c) => ((InstrumentState)c!).Update((double)m, l));
+        }
+
+
+
+        public AggregationManager Include(string meterName)
+        {
+            Include(i => i.Meter.Name == meterName);
+            return this;
+        }
+
+        public AggregationManager Include(string meterName, string instrumentName)
+        {
+            Include(i => i.Meter.Name == meterName && i.Name == instrumentName);
```

Comment:
>i => i.Meter.Name == meterName && i.Name == instrumentName [](http://example.com/codeflow?start=20&length=58)

just wondering, wouldn't this will create a delegate object everytim it gets called?

Reply:
yeah it does. Include only gets called when EventSource is being turned on (or reconfigured) so this happens very rarely overall.